packages/BiostructureViewer/src/utils/ngl-gl-doc-service.ts

```diff
@@ -0,0 +1,210 @@
+import * as grok from 'datagrok-api/grok';
+import * as ui from 'datagrok-api/ui';
+import * as DG from 'datagrok-api/dg';
+
+import {_package} from '../package';
+import * as NGL from 'NGL';
+import {NglGlServiceBase, NglGlTask} from '@datagrok-libraries/bio/src/viewers/ngl-gl-viewer';
+
+
+export class NglGlDocService implements NglGlServiceBase {
+  private readonly nglDiv: HTMLDivElement;
+  private readonly ngl: any;
+
+  private readonly hostDiv: HTMLDivElement | null = null;
+
+  private readonly _queue: { key?: keyof any, task: NglGlTask, dt: number }[];
+  private readonly _queueDict: { [key: keyof any]: NglGlTask };
+  /** The flag allows {@link _processQueue}() on add item to the queue with {@link render}() */
+  private _busy: boolean = false;
+
+  constructor() {
+    const r = window.devicePixelRatio;
+
+    this.nglDiv = ui.div([], 'd4-ngl-viewer');
+    this.ngl = new NGL.Stage(this.nglDiv);
+    this.ngl.viewer.signals.rendered.add(this.onNglRendered.bind(this));
+
+    // The single NGL component
+    this.hostDiv = ui.div([this.nglDiv]);
+    this.hostDiv.style.position = 'absolute';
+    this.hostDiv.style.left = '0px';
+    this.hostDiv.style.right = '0px';
+    this.hostDiv.style.width = '0px';
+    this.hostDiv.style.height = '0px';
+    document.body.appendChild(this.hostDiv);
+
+    this._queue = [];
+    this._queueDict = {};
+
+    window.setInterval(() => { this._sweepQueue(); }, 200);
+  }
+
+  render(task: NglGlTask, key?: keyof any): void {
+    //_package.logger.debug('NglGlService.render() start ' + `name: ${name}`);
+
+    if (key !== undefined) {
+      if (key in this._queueDict) {
+        // remove outdated task from the queue
+        const oldTaskI = this._queue.findIndex((item) => item.key === key);
+        this._queue.splice(oldTaskI, 1);
+      }
+      this._queueDict[key] = task;
+    }
+
+    this._queue.push({key, task, dt: Date.now()});
+
+    if (!this._busy) {
+      this._busy = true;
+
+      // TODO: Use requestAnimationFrame()
+      window.setTimeout(async () => { await this._processQueue(); }, 0 /* next event cycle */);
+    }
+    //_package.logger.debug('NglGlService.render() end ' + `name: ${name}`);
+  }
+
+  private async _processQueue() {
+    const {key, task} = this._queue.shift()!;
+    if (key) delete this._queueDict[key];
+    try {
+      const r = window.devicePixelRatio;
+
+      // TODO: Convert string to Blob once converting PDB string column to Blob
+      const stringBlob = new Blob([task.props.pdb], {type: 'text/plain'});
+
+      this.task = undefined;
```

Comment:
Why are you setting it to undefined if they are reset again on line 93?

Reply:
I use this value as the flag in the event handler onNglRendered() to prevent redundant copying of a white space drawing to the grid canvas (it works unstable, though).